src/shogun/base/SGObject.h

```diff
@@ -452,16 +458,16 @@ class CSGObject
 		      class X = typename std::enable_if<is_sg_base<T>::value>::type>
 	T* get(const std::string& name, index_t index, std::nothrow_t) const
 	{
-		CSGObject* result = nullptr;
+		Derived* result = nullptr;
 
 		auto get_lambda = [&index, &result](auto& array) {
 			result = array.at(index);
 		};
-		if (sgo_details::dispatch_array_type<T>(this, name, get_lambda))
+		if (sgo_details::dispatch_array_type<T>((Derived*) this, name, get_lambda))
```

Comment:
could you do explicit cast with c++ styles? like static_cast<T>()... again later refactor of code will be made easier this way